### ISSUE TITLE
Pyright shows error for overloaded union arguments when using generics

### ISSUE THREAD

**Issue by AlexandrKrivosheev** [microsoft/pyright#1088]
**Describe the bug**
If method argument is Union[X, Y], pyright will say that X cannot be assigned to X|Y

```
Argument of type "TxFetcherCrypto" cannot be assigned to parameter "tx_fetcher" of type "ITxFetcher[TTx]" in function "__init__"
  TypeVar "TTx" is invariant
    Type "TCryptoTx | TFiatTx" cannot be assigned to type "TCryptoTx"
    Type "TCryptoTx" cannot be assigned to type "TTx"
```

**Expected behavior**
No errors shown

**Screenshots or Code**

```python
from typing import Generic, Protocol, TypeVar, TypedDict, Union, overload

TTx = TypeVar("TTx")


class ITxFetcher(Protocol[TTx]):
    def fetch(self) -> TTx:
        ...


class ITxSaver(Protocol[TTx]):
    def save(self, tx: TTx):
        ...


class TxProcessingController(Generic[TTx]):
    def __init__(self, tx_fetcher: ITxFetcher[TTx], tx_saver: ITxSaver[TTx]) -> None:
        tx_saver.save(tx_fetcher.fetch())


class TFiatTx(TypedDict):
    fiat: str


class TCryptoTx(TypedDict):
    crypto: str


class TxFetcherCrypto:
    def fetch(self) -> TCryptoTx:
        return TCryptoTx(crypto="")


class TxFetcherFiat:
    def fetch(self) -> TFiatTx:
        return TFiatTx(fiat="")


class TxSaver:
    @overload
    def save(self, tx: TFiatTx):
        print("save fiat")

    @overload
    def save(self, tx: TCryptoTx):
        print("save crypto")

    def save(self, tx: Union[TCryptoTx, TFiatTx]):
        print("save union")
TxProcessingController(tx_fetcher=TxFetcherCrypto(), tx_saver=TxSaver()) #error
# if overloaded methods in TxSaver is swapped (TCryptoTx is first), 
# error will be shown for second instantiation 
TxProcessingController(tx_fetcher=TxFetcherFiat(), tx_saver=TxSaver())
```

**VS Code extension or command-line**
vscode pylance v2020.9.8

**Comment by erictraut** [comment]
I think Pyright is doing the right thing here. As evidence, mypy also reports errors with the code above.

There are a number of issues I see with this code.

First, the code is abusing overloads in a manner they are not intended to be used. Overloads are meant as a way to specify different return types based on input parameter types. Your overloads all have the same return type. So let's start by deleting the two overloaded methods.

Once the overloads are deleted, you run into the problem that `ITxFetcher` and `ITxSaver` both use invariant type variables. This means the type solver must match the type exactly. No solution exists when solving for the type variables in either of your instantiations of `TxProcessingController`.

I see two solutions here:
1. You can change the type variable used in the declaration for `ITxFetcher` so it is covariant.

```python
TTx_co = TypeVar("TTx_co", covariant=True)

class ITxFetcher(Protocol[TTx_co]):
    def fetch(self) -> TTx_co:
        ...
```

2. You can make `TxSaver` generic parameterized by a constrained type variable:

```python
TTxSaved = TypeVar("TTxSaved", TCryptoTx, TFiatTx)

class TxSaver(Generic[TTxSaved]):
    def save(self, tx: TTxSaved):
        pass
```

**Comment by AlexandrKrivosheev** [comment]
Thank you for explanation 🙏 
I will close the issue.
